Move RPC registration out of AppInitParameterInteraction

ryanofsky opened the pull request:
Move to AppInitServers. This doesn't have any effects on bitcoin behavior. It was just strange to have this unrelated code in the middle of parameter interaction.

ryanofsky committed:
Move RPC registration out of AppInitParameterInteraction

Move to AppInitServers. This doesn't have any effects on bitcoin behavior. It
was just strange to have this unrelated code in the middle or parameter
interaction.

promag commented on src/init.cpp:
```diff
@@ -709,6 +709,10 @@ bool InitSanityCheck(void)
 
 bool AppInitServers(boost::thread_group& threadGroup)
 {
+    RegisterAllCoreRPCCommands(tableRPC);
```
Nit, could be in `OnRPCStarted`, like there is no need to register if it fails to start?

promag reviewed (commented):
Of course ACK.

TheBlueMatt commented:
Does this break using the in-Qt RPC console when running without -server?

promag commented:
@TheBlueMatt is right, with this change executing `help` in the Qt console gives:
```
== Control ==
help ( "command" )
stop
uptime
```

laanwj commented:
> Does this break using the in-Qt RPC console when running without -server?

Whops - it appears the logic is wrong there currently. Qt debug console is considered a front-end for RPC.
So `StartRPC` is definitely supposed to be called when Qt's debug console is active. Same for the `OnRPCStarted `OnRPCStopped` registrations in InitServers, probably.
It's just the HTTP and REST stuff which can be left out if `-server` is not provided.

ryanofsky commented:
Updated ccb2a53d9bf0858e963161b2fcaa76639457b410 -> abbd230217df2f3b0cc25527f3c15f6f11b9852e ([pr/rpcinit.1](https://github.com/ryanofsky/bitcoin/commits/pr/rpcinit.1) -> [pr/rpcinit.2](https://github.com/ryanofsky/bitcoin/commits/pr/rpcinit.2)) to fix the problem with qt.

Added a test for the bug matt noticed in #11625.

@laanwj I can also go ahead and make the StartRPC() call not depend on `-server`, but that's a little more awkward than it might otherwise be because StartRPC is sandwiched between InitHTTPServer and StartHTTPRPC calls. Would it be ok to move it above? Also it seems like the http server is not started at all if `-rest` is disabled. Is that intentional?

laanwj commented:
> Would it be ok to move it above?

Yes - constraint is that StartRPC needs to be called before StartHTTPRPC - the relative order of InitHTTPServer and StartRPC shouldn't matter because the parts don't 'know' each other at that point.

> Also it seems like the http server is not started at all if -rest is disabled. Is that intentional?

I'm fairly sure that's not true in master, we'd have plenty of reports, as `-rest` is disabled by default.

promag commented on src/init.cpp:
```diff
@@ -1241,6 +1236,14 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
     GetMainSignals().RegisterBackgroundSignalScheduler(scheduler);
 
+    /* Register RPC commands regardless of -server setting so they will be
+     * available in the GUI RPC console even if external calls are disabled.
```
Only relevant in bitcoin-qt? No need to register if bitcoind without server.

laanwj replied:
That's very rarely used. It doesn't hurt to register the commands anyway. If that helps keep the code simpler.

promag replied:
Yeah agree.

laanwj commented:
Tested ACK abbd230217df2f3b0cc25527f3c15f6f11b9852e

TheBlueMatt commented:
Post-merge utACK abbd230217df2f3b0cc25527f3c15f6f11b9852e
